lib/translators/TerrainTranslator.js

```diff
@@ -35,37 +35,136 @@ const TerrainTranslator = {
         /*
          * Map size data
          */
-        outBuffer.addInt(terrainJson.map.width);
-        outBuffer.addInt(terrainJson.map.height);
+        outBuffer.addInt(terrainJson.map.width + 1);
+        outBuffer.addInt(terrainJson.map.height + 1);
 
-        // Unsupported
-        // map.offset is not implemented yet.. so we hardcode
-        outBuffer.addByte(0x00);outBuffer.addByte(0x00);outBuffer.addByte(0x80);outBuffer.addByte(0xc5);
-        outBuffer.addByte(0x00);outBuffer.addByte(0x00);outBuffer.addByte(0x80);outBuffer.addByte(0xc5);
-        //outBuffer.addFloat(terrainJson.map.offset.x);
-        //outBuffer.addFloat(terrainJson.map.offset.y);
+        // map offset
+        outBuffer.addFloat(terrainJson.map.offset.x);
+        outBuffer.addFloat(terrainJson.map.offset.y);
 
 
         /*
          * Tile points
          */
-        for(let i = terrainJson.tiles.length - 1; i >= 0; i--) {
+        for(let i = 0; i < terrainJson.tiles.length; i++) {
             // write each for of tiles
-            terrainJson.tiles[i].forEach(function(tile) {
-                outBuffer.addShort(tile[0]);            // height
-                outBuffer.addShort(tile[1]);            // water
-                outBuffer.addByte(tile[2]+''+tile[3]);  // flags/tile
-                outBuffer.addByte(tile[4]);             // texture details
-                outBuffer.addByte(tile[5]+''+tile[6]);  // cliff tile/layer
-            });
+            for (let width = 0; width < terrainJson.tiles[i].length; width++) {
+                let currTile = terrainJson.tiles[i][width];
+
+                let boundaryFlag = currTile.boundaryFlag ? 0x4000 : 0;
+
+                outBuffer.addShort(currTile.groundHeight); // height
+                outBuffer.addShort(currTile.waterHeight | boundaryFlag); // water
+                outBuffer.addByte(currTile.flags | currTile.groundTexture);
+                outBuffer.addByte(currTile.groundVariation | currTile.cliffVariation);
+                outBuffer.addByte(currTile.cliffTexture | currTile.layerHeight);
```

Comment:
This section is much appreciated. Adding names to the current tile like `groundHeight`, `waterHeight`, etc. makes this much more readable and self-explanatory.

Could you explain where those bitwise or's are documented from the .w3x spec? Did I happen to miss those when coding the previous version?

Reply:
great! I thought so, too. I saw your note on the wiki about possibly changing the format of the data, so went ahead and did it.

Yeah, I basically got them from https://github.com/stijnherfst/HiveWE/wiki/war3map.w3e-Terrain and then just converted the binary into decimal. Basically using the ORs to combine the bits from various flags to create the byte. We can update some of the magic numbers in the read function so that they're more meaningful. The 4 bit ones are used twice, and then there's a 5 and 3 bits onw.

Reply:
Looks like better / more up-to-date documentation than what I was referencing.

Could you add that link into the source code in a comment? Adding that would help document where those values come from.